samza-test/src/main/java/org/apache/samza/test/framework/TestRunner.java

```diff
@@ -0,0 +1,343 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.test.framework;
+
+import com.google.common.base.Preconditions;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.samza.application.StreamApplication;
+import org.apache.samza.config.Config;
+import org.apache.samza.config.JobConfig;
+import org.apache.samza.config.JobCoordinatorConfig;
+import org.apache.samza.config.MapConfig;
+import org.apache.samza.config.TaskConfig;
+import org.apache.samza.container.grouper.task.SingleContainerGrouperFactory;
+import org.apache.samza.operators.KV;
+import org.apache.samza.runtime.LocalApplicationRunner;
+import org.apache.samza.standalone.PassthroughCoordinationUtilsFactory;
+import org.apache.samza.standalone.PassthroughJobCoordinatorFactory;
+import org.apache.samza.system.EndOfStreamMessage;
+import org.apache.samza.system.IncomingMessageEnvelope;
+import org.apache.samza.system.OutgoingMessageEnvelope;
+import org.apache.samza.system.StreamSpec;
+import org.apache.samza.system.SystemProducer;
+import org.apache.samza.system.SystemStream;
+import org.apache.samza.system.SystemStreamMetadata;
+import org.apache.samza.system.SystemStreamPartition;
+import org.apache.samza.system.inmemory.InMemorySystemConsumer;
+import org.apache.samza.system.inmemory.InMemorySystemFactory;
+import org.apache.samza.task.AsyncStreamTask;
+import org.apache.samza.task.StreamTask;
+import org.apache.samza.test.framework.stream.CollectionStream;
+import org.apache.samza.test.framework.system.CollectionStreamSystem;
+
+
+/**
+ * TestRunner provides apis to quickly set up tests for Samza low level and high level apis. Default running mode
+ * for test is Single container without any distributed coordination service. Test runner maintains global job config
+ * {@code configs} that are used to run the Samza job
+ *
+ * For single container mode following configs are set by default
+ *  <ol>
+ *    <li>"job.coordination.utils.factory" = {@link PassthroughCoordinationUtilsFactory}</li>
+ *    <li>"job.coordination.factory" = {@link PassthroughJobCoordinatorFactory}</li>
+ *    <li>"task.name.grouper.factory" = {@link SingleContainerGrouperFactory}</li>
+ *    <li>"job.name" = "test-samza"</li>
+ *    <li>"processor.id" = "1"</li>
+ *  </ol>
+ *
+ */
+public class TestRunner {
+
+  private static final String JOB_NAME = "test-samza";
+
+  public enum Mode {
+    SINGLE_CONTAINER, MULTI_CONTAINER
+  }
+
+  private static Map<String, String> configs;
+  private static Map<String, CollectionStreamSystem> systems;
```

Comment:
for api to be fluent ex for Utilities like TestRunner.consumeStream() it needs a static map of systems, @xinyuiscool suggested to use ThreadLocal for it, this enables parallel running tests to have independent state

Reply:
Aren't you already creating a new TestRunner for each test? If so, changing these to be non-static seems like it would give you what you want. Static variables are there to share data across instances, but it doesn't seem like you really need that anyways.